{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The purpose of this notebook is to generate the AF_Cutoff_vs_Gene_Number figure for the manuscript."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Parameters/Arguments\n",
    "\n",
    "working_dir = \"/home/jupyter/notebooks/Ancestry\"\n",
    "workspace_bucket = Sys.getenv('WORKSPACE_BUCKET')\n",
    "out_directory = \"avana14\" #the output directory name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "── \u001b[1mAttaching packages\u001b[22m ─────────────────────────────────────── tidyverse 1.3.1 ──\n",
      "\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2\u001b[39m 3.3.5     \u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.1.2     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.4.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.1.3     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 1.4.0     \n",
      "\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\n",
      "------------------------------------------------------------------------------\n",
      "\n",
      "You have loaded plyr after dplyr - this is likely to cause problems.\n",
      "If you need functions from both plyr and dplyr, please load plyr first, then dplyr:\n",
      "library(plyr); library(dplyr)\n",
      "\n",
      "------------------------------------------------------------------------------\n",
      "\n",
      "\n",
      "Attaching package: ‘plyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    compact\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘reshape2’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    smiths\n",
      "\n",
      "\n",
      "Loading required package: viridisLite\n",
      "\n",
      "\n",
      "Attaching package: ‘scales’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:viridis’:\n",
      "\n",
      "    viridis_pal\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    discard\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:readr’:\n",
      "\n",
      "    col_factor\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Set up the environment\n",
    "\n",
    "#load packages\n",
    "library(dplyr)\n",
    "library(tidyverse)\n",
    "library(stringr)\n",
    "library(plyr)\n",
    "library(ggplot2)\n",
    "library(reshape2)\n",
    "library(RColorBrewer)\n",
    "library(viridis)\n",
    "library(scales)\n",
    "\n",
    "#Define functions\n",
    "show_msg <- function(x){ \n",
    "    print(x)\n",
    "    flush.console()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr1.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr10.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr11.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr12.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr13.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr14.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr15.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr16.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr17.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr18.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr19.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr2.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr20.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr21.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr22.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr3.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr4.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr5.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr6.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr7.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr8.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chr9.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chrX.vcf.bgz.avana14'</li><li>'gnomad.genomes.v3.1.hgdp_1kg_subset.chrY.vcf.bgz.avana14'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr1.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr10.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr11.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr12.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr13.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr14.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr15.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr16.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr17.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr18.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr19.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr2.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr20.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr21.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr22.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr3.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr4.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr5.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr6.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr7.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr8.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chr9.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chrX.vcf.bgz.avana14'\n",
       "\\item 'gnomad.genomes.v3.1.hgdp\\_1kg\\_subset.chrY.vcf.bgz.avana14'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr1.vcf.bgz.avana14'\n",
       "2. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr10.vcf.bgz.avana14'\n",
       "3. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr11.vcf.bgz.avana14'\n",
       "4. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr12.vcf.bgz.avana14'\n",
       "5. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr13.vcf.bgz.avana14'\n",
       "6. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr14.vcf.bgz.avana14'\n",
       "7. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr15.vcf.bgz.avana14'\n",
       "8. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr16.vcf.bgz.avana14'\n",
       "9. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr17.vcf.bgz.avana14'\n",
       "10. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr18.vcf.bgz.avana14'\n",
       "11. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr19.vcf.bgz.avana14'\n",
       "12. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr2.vcf.bgz.avana14'\n",
       "13. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr20.vcf.bgz.avana14'\n",
       "14. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr21.vcf.bgz.avana14'\n",
       "15. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr22.vcf.bgz.avana14'\n",
       "16. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr3.vcf.bgz.avana14'\n",
       "17. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr4.vcf.bgz.avana14'\n",
       "18. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr5.vcf.bgz.avana14'\n",
       "19. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr6.vcf.bgz.avana14'\n",
       "20. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr7.vcf.bgz.avana14'\n",
       "21. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr8.vcf.bgz.avana14'\n",
       "22. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chr9.vcf.bgz.avana14'\n",
       "23. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chrX.vcf.bgz.avana14'\n",
       "24. 'gnomad.genomes.v3.1.hgdp_1kg_subset.chrY.vcf.bgz.avana14'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr1.vcf.bgz.avana14\" \n",
       " [2] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr10.vcf.bgz.avana14\"\n",
       " [3] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr11.vcf.bgz.avana14\"\n",
       " [4] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr12.vcf.bgz.avana14\"\n",
       " [5] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr13.vcf.bgz.avana14\"\n",
       " [6] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr14.vcf.bgz.avana14\"\n",
       " [7] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr15.vcf.bgz.avana14\"\n",
       " [8] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr16.vcf.bgz.avana14\"\n",
       " [9] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr17.vcf.bgz.avana14\"\n",
       "[10] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr18.vcf.bgz.avana14\"\n",
       "[11] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr19.vcf.bgz.avana14\"\n",
       "[12] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr2.vcf.bgz.avana14\" \n",
       "[13] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr20.vcf.bgz.avana14\"\n",
       "[14] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr21.vcf.bgz.avana14\"\n",
       "[15] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr22.vcf.bgz.avana14\"\n",
       "[16] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr3.vcf.bgz.avana14\" \n",
       "[17] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr4.vcf.bgz.avana14\" \n",
       "[18] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr5.vcf.bgz.avana14\" \n",
       "[19] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr6.vcf.bgz.avana14\" \n",
       "[20] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr7.vcf.bgz.avana14\" \n",
       "[21] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr8.vcf.bgz.avana14\" \n",
       "[22] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chr9.vcf.bgz.avana14\" \n",
       "[23] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chrX.vcf.bgz.avana14\" \n",
       "[24] \"gnomad.genomes.v3.1.hgdp_1kg_subset.chrY.vcf.bgz.avana14\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"All of the files are here!\"\n"
     ]
    }
   ],
   "source": [
    "#get a list of all of the vcf files\n",
    "\n",
    "paste({working_dir}, \"/filtered_output/\", out_directory, sep = \"\") %>% setwd()\n",
    "file.list <- list.files(path = \".\", pattern = \"gnomad.genomes\")\n",
    "file.list\n",
    "\n",
    "if(length(file.list) == 24){print(\"All of the files are here!\")} else {print(\"There are either too many or too few files\")}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>chromosome</th><th scope=col>startPosition</th><th scope=col>endPosition</th><th scope=col>sgrna</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>chr1</td><td> 69527</td><td> 69550</td><td>TCTCCATTCGGTGAGCCAGT</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>chr1</td><td>168646</td><td>168669</td><td>TGTAGTCCCAGCTACTCAGG</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>chr1</td><td>383420</td><td>383443</td><td>GAGAATCTCTTGAACCCGGG</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>chr1</td><td>424270</td><td>424293</td><td>TGTAATCCCAGTACTTTGGG</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>chr1</td><td>451271</td><td>451294</td><td>ATAATGGTCAGATAGTGGAG</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>chr1</td><td>451344</td><td>451367</td><td>TCTCCACACCACCAACGACG</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & chromosome & startPosition & endPosition & sgrna\\\\\n",
       "  & <chr> & <int> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & chr1 &  69527 &  69550 & TCTCCATTCGGTGAGCCAGT\\\\\n",
       "\t2 & chr1 & 168646 & 168669 & TGTAGTCCCAGCTACTCAGG\\\\\n",
       "\t3 & chr1 & 383420 & 383443 & GAGAATCTCTTGAACCCGGG\\\\\n",
       "\t4 & chr1 & 424270 & 424293 & TGTAATCCCAGTACTTTGGG\\\\\n",
       "\t5 & chr1 & 451271 & 451294 & ATAATGGTCAGATAGTGGAG\\\\\n",
       "\t6 & chr1 & 451344 & 451367 & TCTCCACACCACCAACGACG\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 4\n",
       "\n",
       "| <!--/--> | chromosome &lt;chr&gt; | startPosition &lt;int&gt; | endPosition &lt;dbl&gt; | sgrna &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 1 | chr1 |  69527 |  69550 | TCTCCATTCGGTGAGCCAGT |\n",
       "| 2 | chr1 | 168646 | 168669 | TGTAGTCCCAGCTACTCAGG |\n",
       "| 3 | chr1 | 383420 | 383443 | GAGAATCTCTTGAACCCGGG |\n",
       "| 4 | chr1 | 424270 | 424293 | TGTAATCCCAGTACTTTGGG |\n",
       "| 5 | chr1 | 451271 | 451294 | ATAATGGTCAGATAGTGGAG |\n",
       "| 6 | chr1 | 451344 | 451367 | TCTCCACACCACCAACGACG |\n",
       "\n"
      ],
      "text/plain": [
       "  chromosome startPosition endPosition sgrna               \n",
       "1 chr1        69527         69550      TCTCCATTCGGTGAGCCAGT\n",
       "2 chr1       168646        168669      TGTAGTCCCAGCTACTCAGG\n",
       "3 chr1       383420        383443      GAGAATCTCTTGAACCCGGG\n",
       "4 chr1       424270        424293      TGTAATCCCAGTACTTTGGG\n",
       "5 chr1       451271        451294      ATAATGGTCAGATAGTGGAG\n",
       "6 chr1       451344        451367      TCTCCACACCACCAACGACG"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#load in the bed file that contains the information for all of the guides\n",
    "\n",
    "paste({working_dir}, \"/filtered_output/\", out_directory, sep = \"\") %>% setwd()\n",
    "bedFile <- read.table('Avana14_filtering.bed', sep = \"\\t\")\n",
    "colnames(bedFile) <- c(\"chromosome\", \"startPosition\", \"endPosition\", \"sgrna\")\n",
    "bedFile <- bedFile[!bedFile$chromosome == \"chrY\",]\n",
    "bedFile <- bedFile[!duplicated(bedFile[ , c(\"sgrna\")]),]\n",
    "head(bedFile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 1\n"
     ]
    }
   ],
   "source": [
    "#Determine which guides beling to each SNP\n",
    "snp.to.guide <- NULL\n",
    "for(file in 1:length(file.list)){\n",
    "show_msg(file)\n",
    "\n",
    "\n",
    "snp.data <- read.table(file.list[file]) #load in the snp data\n",
    "colnames(snp.data) <- c(\"chrom\", \"pos\", \"id\", \"ref\", \"alt\", \"qual\", \"filter\", \"info\") #assign the columns names\n",
    "chromosome.of.interest <- snp.data$chrom %>% unique()\n",
    "bed.file.subset <- bedFile[bedFile$chromosome %in% chromosome.of.interest, ]\n",
    "\n",
    "\t#Find which SNPs map to which guides\n",
    "\tall.kept.snps <- NULL\n",
    "\tfor(i in 1:nrow(bed.file.subset)){\n",
    "\tguideOfInterest <- bed.file.subset[i,] #Extract the row with the sgrna that we are interested in\n",
    "\tkeep.list <- between(snp.data[, 2], guideOfInterest$startPosition, guideOfInterest$endPosition) #Find the SNPs that fall within this sgrna sequence\n",
    "\tkept.snps <- snp.data[keep.list, ] #extract just the snps that we want to keep from the data frame\n",
    "\tguide.sequence <- guideOfInterest$sgrna #get the sequence of the guide for this iteration of the loop\n",
    "\tguide.sequence <- rep(guide.sequence, nrow(kept.snps)) #make a vector so that it can be bound to the df\n",
    "\tkept.snps <- cbind(guide.sequence, kept.snps) #bind them together\n",
    "\tall.kept.snps <- rbind(all.kept.snps, kept.snps)\n",
    "\t}\n",
    "\n",
    "snp.to.guide <- rbind(snp.to.guide, all.kept.snps)\n",
    "}\n",
    "\n",
    "head(snp.to.guide)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Map the guides to genes\n",
    "paste({working_dir}, \"/filtered_output/\", out_directory, sep = \"\") %>% setwd()\n",
    "guide.to.gene <- read.table('avana14_guide_to_gene.chip', sep = \"\\t\", header = T) #load in a matrix that maps genes to each sgrna sequence\n",
    "snp.to.guide <- cbind(snp.to.guide$guide.sequence, snp.to.guide) #duplicate the sgrna column so that we can replace it with the gene name\n",
    "snp.to.guide[,1] <- plyr::mapvalues(snp.to.guide[,1], from = guide.to.gene$Barcode.Sequence, to = guide.to.gene$Gene.Symbol) #Add in the gene name to the dataset so that we know which guide is associated with each gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Write the dataset so that we can use it in the future without needing to re-compute everything\n",
    "paste({working_dir}, \"/filtered_output/\", out_directory, sep = \"\") %>% setwd()\n",
    "write.table(snp.to.guide, \"snp_to_gene_af_matrix.txt\", sep = \"\\t\", col.names = NA, row.names = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract the overall allele frquency data from the snp.to.guide matrix\n",
    "info.vector <- snp.to.guide$info %>% as.vector() #first subset out the info column so that it's a vector, just to make writing the code a little easier.\n",
    "\n",
    "#This is a vetor of all of the phenotype values that we want to extract\n",
    "values.to.extract <- c(\"AC-oth=\", \"AN-oth=\", \n",
    "\t\t\t\t\t\t\"AC-ami=\", \"AN-ami=\", \n",
    "\t\t\t\t\t\t\"AC-sas=\", \"AN-sas=\", \n",
    "\t\t\t\t\t\t\"AC-fin=\", \"AN-fin=\",\n",
    "\t\t\t\t\t\t\"AC-eas=\", \"AN-eas=\",\n",
    "\t\t\t\t\t\t\"AC-amr=\", \"AN-amr=\", \n",
    "\t\t\t\t\t\t\"AC-afr=\", \"AN-afr=\", \n",
    "\t\t\t\t\t\t\"AC-mid=\", \"AN-mid=\", \n",
    "\t\t\t\t\t\t\"AC-asj=\", \"AN-asj=\", \n",
    "\t\t\t\t\t\t\"AC-nfe=\", \"AN-nfe=\")\n",
    "\t\t\t\t\t\t\n",
    "#This vector contains what we want to call all of the extracted values\n",
    "extracted.value.names <- c(\"ac.oth\", \"an.oth\", \n",
    "\t\t\t\t\t\t\"ac.ami\", \"an.ami\", \n",
    "\t\t\t\t\t\t\"ac.sas\", \"an.sas\", \n",
    "\t\t\t\t\t\t\"ac.fin\", \"an.fin\",\n",
    "\t\t\t\t\t\t\"ac.eas\", \"an.eas\",\n",
    "\t\t\t\t\t\t\"ac.amr\", \"an.amr\", \n",
    "\t\t\t\t\t\t\"ac.afr\", \"an.afr\", \n",
    "\t\t\t\t\t\t\"ac.mid\", \"an.mid\", \n",
    "\t\t\t\t\t\t\"ac.asj\", \"an.asj\", \n",
    "\t\t\t\t\t\t\"ac.nfe\", \"an.nfe\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "AC-oth\tAlternate allele count for samples of Other ancestry\n",
    "AN-oth\tTotal number of alleles in samples of Other ancestry\n",
    "AC-ami\tAlternate allele count for samples of Amish ancestry\n",
    "AN-ami\tTotal number of alleles in samples of Amish ancestry\n",
    "AC-sas\tAlternate allele count for samples of South Asian ancestry\n",
    "AN-sas\tTotal number of alleles in samples of South Asian ancestry\n",
    "AC-fin\tAlternate allele count for samples of Finnish ancestry\n",
    "AN-fin\tTotal number of alleles in samples of Finnish ancestry\n",
    "AC-eas\tAlternate allele count for samples of East Asian ancestry\n",
    "AN-eas\tTotal number of alleles in samples of East Asian ancestry\n",
    "AC-amr\tAlternate allele count for samples of Latino ancestry\n",
    "AN-amr\tTotal number of alleles in samples of Latino ancestry\n",
    "AC-afr\tAlternate allele count for samples of African/African-American ancestry\n",
    "AN-afr\tTotal number of alleles in samples of African/African-American ancestry\n",
    "AC-mid\tAlternate allele count for samples of Middle Eastern ancestry\n",
    "AN-mid\tTotal number of alleles in samples of Middle Eastern ancestry\n",
    "AC-asj\tAlternate allele count for samples of Ashkenazi Jewish ancestry\n",
    "AN-asj\tTotal number of alleles in samples of Ashkenazi Jewish ancestry\n",
    "AC-nfe\tAlternate allele count for samples of Non-Finnish European ancestry\n",
    "AN-nfe\tTotal number of alleles in samples of Non-Finnish European ancestry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract/purify the allele frequency data\n",
    "extracted.value.list <- NULL\n",
    "for(i in 1:length(values.to.extract)){\n",
    "print(i)\n",
    "\n",
    "first.filter <- gsub(\n",
    "\tprint(paste(\".*\", values.to.extract[i], sep = \"\")), \"\", info.vector)\n",
    "\n",
    "second.filter <- gsub(\";.*\", \"\", first.filter) %>% as.numeric()\n",
    "\n",
    "extracted.value.list[[i]] <- second.filter\n",
    "\n",
    "}\n",
    "names(extracted.value.list) <- extracted.value.names #assign the names to the list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Bind it all together into a single data frame\n",
    "extracted.value.df <- NULL\n",
    "for(i in 1:length(extracted.value.list)){\n",
    "extracted.value.df <- cbind(extracted.value.df, extracted.value.list[[i]]) %>% data.frame()\n",
    "}\n",
    "colnames(extracted.value.df) <- extracted.value.names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculate the overall allele frequency\n",
    "total.maf.sum <- rowSums(extracted.value.df[,c(1,3,5,7,9,11,13,15,17,19)]) %>% as.vector()\n",
    "total.genotyped.sum <- rowSums(extracted.value.df[,c(2,4,6,8,10,12,14,16,18,20)]) %>% as.vector()\n",
    "extracted.value.df <- cbind(extracted.value.df, total.maf.sum, total.genotyped.sum) #bind the overall allele count data to the main matrix\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create a new matrix that has the allele frequency data\n",
    "af.oth <- extracted.value.df[,1]/extracted.value.df[,2]\n",
    "af.ami <- extracted.value.df[,3]/extracted.value.df[,4]\n",
    "af.sas <- extracted.value.df[,5]/extracted.value.df[,6]\n",
    "af.fin <- extracted.value.df[,7]/extracted.value.df[,8]\n",
    "af.eas <- extracted.value.df[,9]/extracted.value.df[,10]\n",
    "af.amr <- extracted.value.df[,11]/extracted.value.df[,12]\n",
    "af.afr <- extracted.value.df[,13]/extracted.value.df[,14]\n",
    "af.mid <- extracted.value.df[,15]/extracted.value.df[,16]\n",
    "af.asj <- extracted.value.df[,17]/extracted.value.df[,18]\n",
    "af.nfe <- extracted.value.df[,19]/extracted.value.df[,20]\n",
    "af.tot <- extracted.value.df[,21]/extracted.value.df[,22]\n",
    "af.matrix <- cbind(af.oth, af.ami, af.sas, af.fin, af.eas, af.amr, af.afr, af.mid, af.asj, af.nfe, af.tot) %>% data.frame()\n",
    "colnames(af.matrix) <- c(\"af.oth\", \"af.ami\", \"af.sas\", \"af.fin\", \"af.eas\", \"af.amr\", \"af.afr\", \"af.mid\", \"af.asj\", \"af.nfe\", \"af.tot\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#bind the allele frequency data back to the snp/guide matrix\n",
    "snp.and.af <- cbind(af.matrix, snp.to.guide) %>% data.frame()\n",
    "colnames(snp.and.af) <- c(\"af.oth\", \"af.ami\", \"af.sas\", \"af.fin\", \"af.eas\", \"af.amr\", \"af.afr\", \"af.mid\", \"af.asj\", \"af.nfe\", \"af.tot\",\n",
    "\t\t\t\t\t\t\t\"gene\", \"sgrna\", \"chr\", \"pos\", \"id\", \"ref\", \"alt\", \"qual\", \"filter\", \"info\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Find the fraction of genes that are affected at each AF cutoff\n",
    "af.cutoff.sequence <- seq(from = 0, to = 1, by = 0.001)\n",
    "\n",
    "affected.genes.vector.tot <- NULL\n",
    "for(i in 1:length(af.cutoff.sequence)){\n",
    "show_msg(\"Working on tot\")\n",
    "af.cutoff <- af.cutoff.sequence[i] #bring the af cutoff inside of the loop\n",
    "\n",
    "df.subset <- snp.and.af[snp.and.af$af.tot >= af.cutoff, ] #subset the dataframe so that it only includes genes which have a SNP greater than the AF cutoff\n",
    "unique.genes <- df.subset$gene %>% unique() #save this data outside of the loop if you want to have a list of genes that are at each allele frequency fraction\n",
    "length.unique.genes <- length(unique.genes) #get the total number of affected genes\n",
    "affected.genes.vector.tot <- c(affected.genes.vector.tot, length.unique.genes)\n",
    "}\n",
    "\n",
    "affected.genes.vector.oth <- NULL\n",
    "for(i in 1:length(af.cutoff.sequence)){\n",
    "show_msg(\"Working on oth\")\n",
    "af.cutoff <- af.cutoff.sequence[i] #bring the af cutoff inside of the loop\n",
    "\n",
    "df.subset <- snp.and.af[snp.and.af$af.oth >= af.cutoff, ] #subset the dataframe so that it only includes genes which have a SNP greater than the AF cutoff\n",
    "unique.genes <- df.subset$gene %>% unique() #save this data outside of the loop if you want to have a list of genes that are at each allele frequency fraction\n",
    "length.unique.genes <- length(unique.genes) #get the total number of affected genes\n",
    "affected.genes.vector.oth <- c(affected.genes.vector.oth, length.unique.genes)\n",
    "}\n",
    "\n",
    "\n",
    "affected.genes.vector.ami <- NULL\n",
    "for(i in 1:length(af.cutoff.sequence)){\n",
    "show_msg(\"Working on ami\")\n",
    "af.cutoff <- af.cutoff.sequence[i] #bring the af cutoff inside of the loop\n",
    "\n",
    "df.subset <- snp.and.af[snp.and.af$af.ami >= af.cutoff, ] #subset the dataframe so that it only includes genes which have a SNP greater than the AF cutoff\n",
    "unique.genes <- df.subset$gene %>% unique() #save this data outside of the loop if you want to have a list of genes that are at each allele frequency fraction\n",
    "length.unique.genes <- length(unique.genes) #get the total number of affected genes\n",
    "affected.genes.vector.ami <- c(affected.genes.vector.ami, length.unique.genes)\n",
    "}\n",
    "\n",
    "\n",
    "affected.genes.vector.sas <- NULL\n",
    "for(i in 1:length(af.cutoff.sequence)){\n",
    "show_msg(\"Working on sas\")\n",
    "af.cutoff <- af.cutoff.sequence[i] #bring the af cutoff inside of the loop\n",
    "\n",
    "df.subset <- snp.and.af[snp.and.af$af.sas >= af.cutoff, ] #subset the dataframe so that it only includes genes which have a SNP greater than the AF cutoff\n",
    "unique.genes <- df.subset$gene %>% unique() #save this data outside of the loop if you want to have a list of genes that are at each allele frequency fraction\n",
    "length.unique.genes <- length(unique.genes) #get the total number of affected genes\n",
    "affected.genes.vector.sas <- c(affected.genes.vector.sas, length.unique.genes)\n",
    "}\n",
    "\n",
    "\n",
    "affected.genes.vector.fin <- NULL\n",
    "for(i in 1:length(af.cutoff.sequence)){\n",
    "show_msg(\"Working on fin\")\n",
    "af.cutoff <- af.cutoff.sequence[i] #bring the af cutoff inside of the loop\n",
    "\n",
    "df.subset <- snp.and.af[snp.and.af$af.fin >= af.cutoff, ] #subset the dataframe so that it only includes genes which have a SNP greater than the AF cutoff\n",
    "unique.genes <- df.subset$gene %>% unique() #save this data outside of the loop if you want to have a list of genes that are at each allele frequency fraction\n",
    "length.unique.genes <- length(unique.genes) #get the total number of affected genes\n",
    "affected.genes.vector.fin <- c(affected.genes.vector.fin, length.unique.genes)\n",
    "}\n",
    "\n",
    "\n",
    "affected.genes.vector.eas <- NULL\n",
    "for(i in 1:length(af.cutoff.sequence)){\n",
    "show_msg(\"Working on eas\")\n",
    "af.cutoff <- af.cutoff.sequence[i] #bring the af cutoff inside of the loop\n",
    "\n",
    "df.subset <- snp.and.af[snp.and.af$af.eas >= af.cutoff, ] #subset the dataframe so that it only includes genes which have a SNP greater than the AF cutoff\n",
    "unique.genes <- df.subset$gene %>% unique() #save this data outside of the loop if you want to have a list of genes that are at each allele frequency fraction\n",
    "length.unique.genes <- length(unique.genes) #get the total number of affected genes\n",
    "affected.genes.vector.eas <- c(affected.genes.vector.eas, length.unique.genes)\n",
    "}\n",
    "\n",
    "\n",
    "affected.genes.vector.amr <- NULL\n",
    "for(i in 1:length(af.cutoff.sequence)){\n",
    "show_msg(\"Working on amr\")\n",
    "af.cutoff <- af.cutoff.sequence[i] #bring the af cutoff inside of the loop\n",
    "\n",
    "df.subset <- snp.and.af[snp.and.af$af.amr >= af.cutoff, ] #subset the dataframe so that it only includes genes which have a SNP greater than the AF cutoff\n",
    "unique.genes <- df.subset$gene %>% unique() #save this data outside of the loop if you want to have a list of genes that are at each allele frequency fraction\n",
    "length.unique.genes <- length(unique.genes) #get the total number of affected genes\n",
    "affected.genes.vector.amr <- c(affected.genes.vector.amr, length.unique.genes)\n",
    "}\n",
    "\n",
    "\n",
    "affected.genes.vector.afr <- NULL\n",
    "for(i in 1:length(af.cutoff.sequence)){\n",
    "show_msg(\"Working on afr\")\n",
    "af.cutoff <- af.cutoff.sequence[i] #bring the af cutoff inside of the loop\n",
    "\n",
    "df.subset <- snp.and.af[snp.and.af$af.afr >= af.cutoff, ] #subset the dataframe so that it only includes genes which have a SNP greater than the AF cutoff\n",
    "unique.genes <- df.subset$gene %>% unique() #save this data outside of the loop if you want to have a list of genes that are at each allele frequency fraction\n",
    "length.unique.genes <- length(unique.genes) #get the total number of affected genes\n",
    "affected.genes.vector.afr <- c(affected.genes.vector.afr, length.unique.genes)\n",
    "}\n",
    "\n",
    "\n",
    "affected.genes.vector.mid <- NULL\n",
    "for(i in 1:length(af.cutoff.sequence)){\n",
    "show_msg(\"Working on mid\")\n",
    "af.cutoff <- af.cutoff.sequence[i] #bring the af cutoff inside of the loop\n",
    "\n",
    "df.subset <- snp.and.af[snp.and.af$af.mid >= af.cutoff, ] #subset the dataframe so that it only includes genes which have a SNP greater than the AF cutoff\n",
    "unique.genes <- df.subset$gene %>% unique() #save this data outside of the loop if you want to have a list of genes that are at each allele frequency fraction\n",
    "length.unique.genes <- length(unique.genes) #get the total number of affected genes\n",
    "affected.genes.vector.mid <- c(affected.genes.vector.mid, length.unique.genes)\n",
    "}\n",
    "\n",
    "\n",
    "affected.genes.vector.asj <- NULL\n",
    "for(i in 1:length(af.cutoff.sequence)){\n",
    "show_msg(\"Working on asj\")\n",
    "af.cutoff <- af.cutoff.sequence[i] #bring the af cutoff inside of the loop\n",
    "\n",
    "df.subset <- snp.and.af[snp.and.af$af.asj >= af.cutoff, ] #subset the dataframe so that it only includes genes which have a SNP greater than the AF cutoff\n",
    "unique.genes <- df.subset$gene %>% unique() #save this data outside of the loop if you want to have a list of genes that are at each allele frequency fraction\n",
    "length.unique.genes <- length(unique.genes) #get the total number of affected genes\n",
    "affected.genes.vector.asj <- c(affected.genes.vector.asj, length.unique.genes)\n",
    "}\n",
    "\n",
    "\n",
    "affected.genes.vector.nfe <- NULL\n",
    "for(i in 1:length(af.cutoff.sequence)){\n",
    "show_msg(\"Working on nfe\")\n",
    "af.cutoff <- af.cutoff.sequence[i] #bring the af cutoff inside of the loop\n",
    "\n",
    "df.subset <- snp.and.af[snp.and.af$af.nfe >= af.cutoff, ] #subset the dataframe so that it only includes genes which have a SNP greater than the AF cutoff\n",
    "unique.genes <- df.subset$gene %>% unique() #save this data outside of the loop if you want to have a list of genes that are at each allele frequency fraction\n",
    "length.unique.genes <- length(unique.genes) #get the total number of affected genes\n",
    "affected.genes.vector.nfe <- c(affected.genes.vector.nfe, length.unique.genes)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#bind the output together in a single data frame so that we can create a plot\n",
    "dataframe.for.plotting <- cbind(af.cutoff.sequence, \n",
    "\t\t\t\t\t\t\t\taffected.genes.vector.afr,\n",
    "\t\t\t\t\t\t\t\taffected.genes.vector.nfe,\n",
    "\t\t\t\t\t\t\t\taffected.genes.vector.ami,\n",
    "\t\t\t\t\t\t\t\taffected.genes.vector.sas,\n",
    "\t\t\t\t\t\t\t\taffected.genes.vector.fin,\n",
    "\t\t\t\t\t\t\t\taffected.genes.vector.eas,\n",
    "\t\t\t\t\t\t\t\taffected.genes.vector.amr,\n",
    "\t\t\t\t\t\t\t\taffected.genes.vector.asj,\n",
    "\t\t\t\t\t\t\t\taffected.genes.vector.mid,\n",
    "\t\t\t\t\t\t\t\taffected.genes.vector.oth,\n",
    "\t\t\t\t\t\t\t\taffected.genes.vector.tot) %>% data.frame()\n",
    "\t\t\t\t\t\t\t\t\n",
    "colnames(dataframe.for.plotting) <- c(\"af_cutoff\", \n",
    "\t\t\t\t\t\t\t\t\t\"AFR\",\n",
    "\t\t\t\t\t\t\t\t\t\"NFE\",\n",
    "\t\t\t\t\t\t\t\t\t\"AMI\",\n",
    "\t\t\t\t\t\t\t\t\t\"SAS\",\n",
    "\t\t\t\t\t\t\t\t\t\"FIN\",\n",
    "\t\t\t\t\t\t\t\t\t\"EAS\",\n",
    "\t\t\t\t\t\t\t\t\t\"AMR\",\n",
    "\t\t\t\t\t\t\t\t\t\"ASJ\",\n",
    "\t\t\t\t\t\t\t\t\t\"MID\",\n",
    "\t\t\t\t\t\t\t\t\t\"OTH\",\n",
    "\t\t\t\t\t\t\t\t\t\"Total\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#melt the df\n",
    "melted.dataframe.for.plotting <- melt(dataframe.for.plotting, id = \"af_cutoff\")\n",
    "colnames(melted.dataframe.for.plotting) <- c(\"af_cutoff\", \"Ancestry\", \"value\")\n",
    "\n",
    "\n",
    "#generate the plot\n",
    "ggplot(melted.dataframe.for.plotting, aes(x = af_cutoff, y = value, color = Ancestry)) +\n",
    "\tgeom_point() +\n",
    "\ttheme_bw() +\n",
    "\tscale_color_brewer(palette = \"Spectral\") +\n",
    "\n",
    "theme(panel.border = element_blank(), panel.grid.major = element_blank(),\n",
    "panel.grid.minor = element_blank(), axis.line = element_line(colour = \"black\")) +\n",
    "theme(axis.text=element_text(size=12),\n",
    "        axis.title=element_text(size=14,face=\"bold\")) +\n",
    "theme(axis.text.x = element_text(face=\"bold\", \n",
    "                           size=14),\n",
    "          axis.text.y = element_text(face=\"bold\", \n",
    "                           size=14)) +\n",
    "\n",
    "theme(\n",
    "panel.border = element_blank(), panel.grid.major = element_blank(),\n",
    "panel.grid.minor = element_blank(), axis.line = element_line(color = \"black\")\n",
    ") +\n",
    "\n",
    "theme(\n",
    "axis.title.x = element_text(color = \"black\", size = 14, face = \"bold\"),\n",
    "axis.title.y = element_text(color = \"black\", size = 14, face = \"bold\")\n",
    ") +\n",
    "\n",
    "theme(\n",
    "axis.text=element_text(size = 12, face = \"bold\"),\n",
    "axis.title=element_text(size = 14,face = \"bold\")\n",
    ") +\n",
    "\n",
    "theme(\n",
    "  legend.title = element_text(size = 14, face = \"bold\"),\n",
    "  legend.text = element_text(size = 10, face = \"bold\")\n",
    ") +\n",
    "\n",
    "xlab(\"Genotype Frequency\") +\n",
    "ylab(\"# Affected Genes\") +\n",
    "\n",
    "\n",
    "scale_y_continuous(breaks=c(0, 1, 10, 100, 1000, 10000), trans=log10_trans())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Export the plot\n",
    "\n",
    "paste(working_dir, \"/figures\", sep =\"\") %>% setwd()\n",
    "ggsave(\"af_cutoff_vs_gene_number.tiff\") #save the plot\n",
    "\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/figures\n",
    "gsutil cp af_cutoff_vs_gene_number.tiff gs://fc-45c0e148-0b1c-4244-9bfc-feb559bbc514\n",
    "\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###This may have errored somewhere in the process. As a next step, check to make sure the figure was output correctly."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.0"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
