{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The purpose of this notebook is to phase genotypes using Eagle. The reason we want to do this is because we want to run RFMix to impute the ancestry fraction of unknown cell lines."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set up the environment and install all of the software"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Arguments/Parameters\n",
    "\n",
    "working_dir = \"/home/jupyter/notebooks/Ancestry\"\n",
    "workspace_bucket = Sys.getenv('WORKSPACE_BUCKET')\n",
    "out_directory = \"avana14\" #the output directory name\n",
    "num.threads = 92 #This is the number of threads that will be used to phase the genotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "── \u001b[1mAttaching packages\u001b[22m ─────────────────────────────────────── tidyverse 1.3.1 ──\n",
      "\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2\u001b[39m 3.3.5     \u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.1.4     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.4.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.1.3     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 2.0.1     \n",
      "\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\n",
      "------------------------------------------------------------------------------\n",
      "\n",
      "You have loaded plyr after dplyr - this is likely to cause problems.\n",
      "If you need functions from both plyr and dplyr, please load plyr first, then dplyr:\n",
      "library(plyr); library(dplyr)\n",
      "\n",
      "------------------------------------------------------------------------------\n",
      "\n",
      "\n",
      "Attaching package: ‘plyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    compact\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘reshape2’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    smiths\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Set up the environment\n",
    "\n",
    "#load packages\n",
    "library(dplyr)\n",
    "library(tidyverse)\n",
    "library(stringr)\n",
    "library(plyr)\n",
    "library(reshape2)\n",
    "\n",
    "#Define functions\n",
    "show_msg <- function(x){ \n",
    "    print(x)\n",
    "    flush.console()\n",
    "}\n",
    "\n",
    "#Make directories\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}\n",
    "\n",
    "#Create the directory to store the liftover files\n",
    "if [ ! -d '{working_dir}/LiftOver' ] \n",
    "then\n",
    "mkdir LiftOver\n",
    "fi\n",
    "\n",
    "if [ ! -d '{working_dir}/LiftOver/hg38_to_hg19' ] \n",
    "then\n",
    "mkdir LiftOver/hg38_to_hg19\n",
    "fi\n",
    "\n",
    "if [ ! -d '{working_dir}/Eagle_Phasing' ] \n",
    "then\n",
    "mkdir {working_dir}/Eagle_Phasing\n",
    "fi\n",
    "\n",
    "if [ ! -d '{working_dir}/rfmix_output' ]\n",
    "then\n",
    "mkdir {working_dir}/rfmix_output\n",
    "fi\n",
    "\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download and install software\n",
    "\n",
    "#Download Eagle v2.4.1\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/software\n",
    "wget https://storage.googleapis.com/broad-alkesgroup-public/Eagle/downloads/Eagle_v2.4.1.tar.gz\n",
    "tar -xf Eagle_v2.4.1.tar.gz\n",
    "rm -rf Eagle_v2.4.1.tar.gz\n",
    "\n",
    "cd {working_dir}/software/Eagle_v2.4.1/tables\n",
    "bgzip -d *.gz\n",
    "\"))\n",
    "\n",
    "\n",
    "#Download and install Samtools\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/software\n",
    "wget https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2\n",
    "tar -vxjf samtools-1.9.tar.bz2\n",
    "cd samtools-1.9\n",
    "make\n",
    "cd {working_dir}/software\n",
    "rm samtools-1.9.tar.bz2\n",
    "\"))\n",
    "\n",
    "#Download and install Bcftools\n",
    "\n",
    "\n",
    "#Set the bcftools plugin path and add bcftools to PATH\n",
    "system(glue::glue(\"\n",
    "export PATH=$PATH:{working_dir}/software/bcftools\n",
    "export BCFTOOLS_PLUGINS='/home/jupyter/notebooks/Ancestry/software/bcftools/plugins'\n",
    "\"))\n",
    "\n",
    "\n",
    "#Download and install tabix\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Install RFMix and process all of the dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download and instal RFMix\n",
    "#I first installed RFMixv2 on the UGER cluster, then zipped and transferred the directory over to the google bucket for this project\n",
    "\n",
    "\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/software\n",
    "\n",
    "#Create the directory to store the rfmix files\n",
    "if [ ! -d '{working_dir}/software/rfmix' ] \n",
    "then\n",
    "mkdir {working_dir}/software/rfmix\n",
    "cd {working_dir}/software/rfmix\n",
    "\n",
    "#Download the rfmix script\n",
    "gsutil cp gs://fc-45c0e148-0b1c-4244-9bfc-feb559bbc514/rfmix.zip .\n",
    "unzip rfmix.zip\n",
    "rm rfmix.zip\n",
    "\n",
    "\"))\n",
    "\n",
    "\n",
    "#Download the reference panel\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/software/rfmix\n",
    "wget ftp://yunlianon:anon@rc-ns-ftp.its.unc.edu/ALL.phase3_v5.shapeit2_mvncall_integrated.noSingleton.tgz\n",
    "tar zxvf ALL.phase3_v5.shapeit2_mvncall_integrated.noSingleton.tgz\n",
    "git clone https://github.com/joepickrell/1000-genomes-genetic-maps.git\n",
    "fi\n",
    "\"))\n",
    "\n",
    "#Download the genetic map\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/software/rfmix\n",
    "git clone https://github.com/joepickrell/1000-genomes-genetic-maps.git\n",
    "\"))\n",
    "\n",
    "\n",
    "#Format the reference panel\n",
    "arg.chunk <- paste('$1 ~ /^#/ {print $0;next} {print $0 | \"sort -k1,1 -k2,2n\"}') #make a variable to insert into the system() function so that the syntax works correctly\n",
    "\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/software/rfmix/all\n",
    "\n",
    "#Index all of the files\n",
    "for vcffile in *.vcf.gz\n",
    "do\n",
    "{working_dir}/software/tabix-0.2.6/tabix -p vcf $vcffile\n",
    "done\n",
    "\n",
    "{working_dir}/software/bcftools/bcftools merge *.vcf.gz --force-samples -o 1kg_phase3_v5_reference.vcf\n",
    "cat 1kg_phase3_v5_reference.vcf | awk {arg.chunk} > 1kg_phase3_v5_reference.sorted.vcfbgzip 1kg_phase3_v5_reference.sorted.vcf\n",
    "{working_dir}/software/tabix-0.2.6/tabix -p vcf 1kg_phase3_v5_reference.sorted.vcf.gz\n",
    "mv 1kg_phase3_v5_reference.sorted.vcf.gz 1kg_phase3_v5_reference.vcf.gz\n",
    "\n",
    "\"))\n",
    "\n",
    "\n",
    "#Format the genetic map\n",
    "paste(working_dir, \"/software/rfmix/1000-genomes-genetic-maps/interpolated_from_hapmap\", sep = \"\") %>% setwd()\n",
    "fileList <- list.files(pattern = \"interpolated\")\n",
    "\n",
    "for(i in 1:length(fileList)){\n",
    "file <- read.table(fileList[i], header = T)\n",
    "Chromosome <- rep(i, nrow(file))\n",
    "output <- cbind(Chromosome, file[ ,2:3])\n",
    "colnames(output) <- c(\"Chromosome\", \"Physical_Position\", \"Genetic_Position\")\n",
    "\n",
    "write.table(output, \n",
    "print(paste(\"chr\", i, \"_genetic_map\", sep = \"\")),\n",
    "quote = FALSE,\n",
    "sep = \"\\t\",\n",
    "row.names = FALSE)\n",
    "}\n",
    "\n",
    "#Download the sample map\n",
    "#the sample map is a two column text file. the first column is the sample names. the second column is the ancestry group.\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/software/rfmix\n",
    "gsutil cp gs://fc-45c0e148-0b1c-4244-9bfc-feb559bbc514/1kg_sample_map .\n",
    "\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Download and process the 1000 genomes reference panel "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download and process the 1000 genomes phase 3 reference panel using a slightly modified version of the code provided in the Eagle v2.4.1 User Manual\n",
    "(https://alkesgroup.broadinstitute.org/Eagle/#x1-250005.1.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#I've already downloaded the 1000 genomes data and put it in the RFMix folder.\n",
    "#Copy it to the Eagle_Phasing directory.\n",
    "\n",
    "system(glue::glue(\"\n",
    "cp {working_dir}/Ancestry/software/rfmix/ALL.phase3_v5.shapeit2_mvncall_integrated.noSingleton.tgz /home/jupyter/notebooks/Ancestry/Eagle_Phasing\n",
    "tar zxvf ALL.phase3_v5.shapeit2_mvncall_integrated.noSingleton.tgz\n",
    "mv all 1000_genomes_vcfs\n",
    "\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download the genome fasta and index it\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing/1000_genomes_vcfs\n",
    "wget -O- http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz | gzip -d > human_g1k_v37.fasta  \n",
    "{working_dir}/software/samtools faidx human_g1k_v37.fasta \n",
    "\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Now do a bunch of data processing for the 1000 genomes vfcs\n",
    "chromosomes = c(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, \"X\", \"Y\")\n",
    "for(chr in chromosomes){\n",
    "    \n",
    "    system(glue::glue(\"\n",
    "    cd {working_dir}/Eagle_Phasing/1000_genomes_vcfs\n",
    "     {working_dir}/software/bcftools/bcftools view --no-version -Ou -c 2 ALL.chr{chr}.phase3_v5.shapeit2_mvncall_integrated.noSingleton.genotypes.vcf.gz |  {working_dir}/software/bcftools/bcftools norm --no-version -Ou -m -any |  {working_dir}/software/bcftools/bcftools norm --no-version -Ob -o ALL.chr{chr}.phase3_integrated.20130502.genotypes.bcf -d none -f human_g1k_v37.fasta && {working_dir}/software/bcftools/bcftools index -f ALL.chr{chr}.phase3_integrated.20130502.genotypes.bcf  \n",
    "    \"))\n",
    "    \n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Download and format the CCLE sequencing data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download the CCLE variant calls and index the vcf."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download the CCLE data\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "gsutil cp gs://fc-45c0e148-0b1c-4244-9bfc-feb559bbc514/ccle.all.called.vcf.gz .\n",
    "\"))\n",
    "\n",
    "#Index ccle.all.called.vcf\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "{working_dir}/software/tabix-0.2.6/tabix -p vcf ccle.all.called.vcf.gz\n",
    "\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The header of the vcf file has some strange cell line names as the sample names. But we don't want this. The CDS names are stored in the vcf header, so we can extract them and then assign them as the sample names in the vcf."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1918</li><li>1</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1918\n",
       "\\item 1\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1918\n",
       "2. 1\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 1918    1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Fix the header in ccle.all.called.vcf to conver the sample ID to the CDS ID\n",
    "\n",
    "#First export the header and format it so that it is ready for us to work with it in R\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "{working_dir}/software/bcftools/bcftools view -h ccle.all.called.vcf.gz -o sample.header.txt\n",
    "head -n 118 sample.header.txt | tail -n 1 | tr -d '#' > sample.header.for.r.txt\n",
    "rm sample.header.txt\n",
    "\"))\n",
    "\n",
    "\n",
    "#Format a dataset where each new \"CDS ID\" sample header is on a new row, then write it.\n",
    "paste(working_dir, \"/Eagle_Phasing\", sep = \"\") %>% setwd()\n",
    "sample.header <- read.table('sample.header.for.r.txt', sep = \"\\t\")\n",
    "sample.header <- sample.header[,1] %>% as.vector()\n",
    "split.row <- data.frame(strsplit(sample.header, \" \")) #split the row on white space\n",
    "split.row <- as.vector(split.row[,1]) #Convert it to a vector\n",
    "split.row <- split.row[9:length(split.row)] #Remove the first 8 elements since they don't contain sample names\n",
    "split.row <- gsub(\".*/\", \"\", split.row) #Each element has a '/' before the CDS, and no slashes after. So we can use gsub to remove all the junk\n",
    "split.row <- gsub('_cnn_filtered.vcf.gz', '', split.row) #Now each CDS ID has \"_cnn_filtered.vcf.gz\" after it, so remove that.\n",
    "split.row <- head(split.row, -5) #Remove the last 5 elelemnts since they are not sample names\n",
    "split.row <- gsub(\";\", \"\", split.row)\n",
    "split.row <- data.frame(split.row)\n",
    "dim(split.row)\n",
    "write.table(split.row, \"cds.name.list.txt\", sep = \"\\t\", col.names = F, row.names = F, quote = F)\n",
    "\n",
    "#Replace the sample names in ccle.all.called.vcf.gz file\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "{working_dir}/software/bcftools/bcftools reheader --samples cds.name.list.txt -o cdsnames.ccle.all.called.vcf.gz ccle.all.called.vcf.gz\n",
    "mv cdsnames.ccle.all.called.vcf.gz ccle.all.called.vcf.gz\n",
    "\"))\n",
    "\n",
    "#Clean up\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "rm cds.name.list.txt\n",
    "rm sample.header.for.r.txt\n",
    "rm ccle.all.called.vcf.gz.tbi\n",
    "\"))\n",
    "\n",
    "#Re-index the vcf file\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "{working_dir}/software/tabix-0.2.6/tabix -p vcf ccle.all.called.vcf.gz\n",
    "\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are some combined genotype calls in the CCLE data where multiple genotypes at the same site are combined on the same line.\n",
    "\n",
    "For example: chr1 34578 A,G,C T.\n",
    "\n",
    "So split it so that it now looks like this:\n",
    "\n",
    "chr1 34578 A T\n",
    "chr1 34578 G T\n",
    "chr1 34578 C T\n",
    "\n",
    "Then re-index the file yet again."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Splitting Genotype Calls\"\n"
     ]
    }
   ],
   "source": [
    "#split the genotype calls\n",
    "show_msg(\"Splitting Genotype Calls\")\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "{working_dir}/software/bcftools/bcftools norm -m - ccle.all.called.vcf.gz -o split.ccle.all.called.vcf.gz\n",
    "mv split.ccle.all.called.vcf.gz ccle.all.called.vcf.gz\n",
    "\"))\n",
    "\n",
    "#clean up\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "rm ccle.all.called.vcf.gz.tbi\n",
    "\"))\n",
    "\n",
    "#Re-index the vcf file\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "{working_dir}/software/tabix-0.2.6/tabix -p vcf ccle.all.called.vcf.gz\n",
    "\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The sequencing data is a mix of WES and WGS data. We don't want to introduce bias because the WGS samples are sequenced better than the WES samples. \n",
    "\n",
    "As a way to try to get around this, we can filter the data so that it only includes variants in exons. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download a bed file that has the exon positions\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "gsutil cp gs://fc-45c0e148-0b1c-4244-9bfc-feb559bbc514/hg19_exon_positions.txt .\n",
    "\"))\n",
    "\n",
    "#Filter the file so that it only includes exons\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "{working_dir}/software/bcftools/bcftools view --force-samples -R hg19_exon_positions.txt ccle.all.called.vcf.gz -o exon.ccle.all.called.vcf.gz -Oz\n",
    "\"))\n",
    "\n",
    "#Clean up\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "rm hg19_exon_positions.txt\n",
    "\"))\n",
    "\n",
    "#Re-index the vcf file\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "{working_dir}/software/tabix-0.2.6/tabix -p vcf exon.ccle.all.called.vcf.gz\n",
    "\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The VCF file has very many samples in it, and not all of them are in DepMap. Filter the vcf so that it only includes the WES and WGS samples that are in DepMap."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download the ccle.sample.tracker from Jeremie/Javad which has information for all of the ccle cell lines\n",
    "#This particular version of the file is stripped to only include the ach id, cds id, and the sequencing platform\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "gsutil cp gs://fc-45c0e148-0b1c-4244-9bfc-feb559bbc514/ccle_sample_tracker.csv .\n",
    "\"))\n",
    "\n",
    "#Download the chronos data for 21Q3Public\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "gsutil cp gs://fc-45c0e148-0b1c-4244-9bfc-feb559bbc514/CRISPR_gene_effect.csv .\n",
    "\"))\n",
    "\n",
    "#Load both datasets into R, then create a list of samples that we want to keep for our analysis.\n",
    "paste(working_dir, \"/Eagle_Phasing\", sep = \"\") %>% setwd()\n",
    "sample.tracker = read.table('ccle_sample_tracker.csv', sep = \",\", header = T)\n",
    "chronos.scores = read.table('CRISPR_gene_effect.csv', sep = \",\", header = T)\n",
    "\n",
    "#Get a list of all of the CDS IDs that we want to keep for our analysis\n",
    "ids.to.keep = sample.tracker %>%\n",
    "filter(datatype %in% c(\"wes\", \"wgs\")) %>%\n",
    "filter(arxspan_id %in% chronos.scores$DepMap_ID) %>%\n",
    "distinct(arxspan_id, .keep_all = TRUE) %>%\n",
    "select(cds_id)\n",
    "ids.to.keep = ids.to.keep[,1] %>% as.vector() %>% unique()\n",
    "\n",
    "#Write them to the disk\n",
    "paste(working_dir, \"/Eagle_Phasing\", sep = \"\") %>% setwd()\n",
    "write.table(ids.to.keep, \"samples.to.keep.txt\", sep = \"\\t\", col.names = FALSE, row.names = FALSE, quote = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Now filter the ccle vcf file so that it only includes the DepMap samples\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "{working_dir}/software/bcftools/bcftools view -S samples.to.keep.txt --force-samples exon.ccle.all.called.vcf.gz -o depmap.exon.ccle.all.called.vcf.gz -Oz\n",
    "\"))\n",
    "\n",
    "#Clean up\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "rm samples.to.keep.txt\n",
    "rm CRISPR_gene_effect.csv\n",
    "rm ccle_sample_tracker.csv\n",
    "\"))\n",
    "\n",
    "#Index the new file\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "{working_dir}/software/tabix-0.2.6/tabix -p vcf depmap.exon.ccle.all.called.vcf.gz\n",
    "\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since we removed some samples, the dataset now has many SNVs that have few (or zero) samples with the SNV in it. Filter the vcf file so that it only includes samples with variants that are at a frequency above 1/1000 to take care of this problem.\n",
    "\n",
    "We don't want to filter too strictly here since SNVs that are at a high frequency in minority populations will still be a low frequency in the DepMap dataset. So for example, if we filtered at a 5% cutoff we would remove a majority of the SNVs in the AFR cell lines simply because they are poorly represented in the DepMap dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter by maf\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "{working_dir}/software/bcftools/bcftools view -i 'MAF > 0.001' depmap.exon.ccle.all.called.vcf.gz -Oz -o maf.ccle.all.called.vcf.gz\n",
    "\"))\n",
    "\n",
    "#index the new vcf that was filtered by maf frequency\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "{working_dir}/software/tabix-0.2.6/tabix -p vcf maf.ccle.all.called.vcf.gz\n",
    "\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The variants in the vcf file are coded with this notation:\n",
    "./.:.:.:.:.\n",
    "\n",
    "Fix it so that these variants are coded as:\n",
    "0/0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "#recode the ./. variants to 0/0\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "{working_dir}/software/bcftools/bcftools +setGT maf.ccle.all.called.vcf.gz -Oz -o recoded.ccle.all.called.vcf.gz -- -t . -n 0\n",
    "\"))\n",
    "\n",
    "#index the new recoded file\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "{working_dir}/software/tabix-0.2.6/tabix -p vcf recoded.ccle.all.called.vcf.gz\n",
    "\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have completed all of the filtering steps, convert the vcf file so that it is a bcf file. This will just help the computing run faster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "#convert recoded.ccle.all.called.vcf.gz to a bcf file\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "{working_dir}/software/bcftools/bcftools convert recoded.ccle.all.called.vcf.gz -Ou -o recoded.ccle.all.called.bcf\n",
    "\"))\n",
    "\n",
    "#then index the bcf file\n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "{working_dir}/software/bcftools/bcftools index -f recoded.ccle.all.called.bcf\n",
    "\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Split the ccle data into individual chromosomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Split the bcf file into individual chromosomes\n",
    "chromosomes = seq(from = 22, to = 1, by = -1)\n",
    "\n",
    "for(chr in chromosomes){\n",
    "    \n",
    "    system(glue::glue(\"\n",
    "    cd {working_dir}/Eagle_Phasing\n",
    "    {working_dir}/software/bcftools/bcftools view -r {chr} recoded.ccle.all.called.bcf -Ou -o chr{chr}.recoded.ccle.all.called.bcf\n",
    "\"))\n",
    "    \n",
    "}\n",
    "\n",
    "\n",
    "#Then index the new files\n",
    "chromosomes = seq(from = 22, to = 1, by = -1)\n",
    "\n",
    "for(chr in chromosomes){\n",
    "    \n",
    "    system(glue::glue(\"\n",
    "    cd {working_dir}/Eagle_Phasing\n",
    "    {working_dir}/software/bcftools/bcftools index -f chr{chr}.recoded.ccle.all.called.bcf\n",
    "\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate the intersection between the 1kg bcf files and the DepMap bcf files so that we can get the SNPs that intersect between the two files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get a list of all of the autosomes\n",
    "chromosomes = seq(from = 22, to = 1, by = -1)\n",
    "\n",
    "#Loop through all of the chromosomes and calculate the intersection between the 1kg bcf and the ccle bcf\n",
    "\n",
    "for(chr in chromosomes){\n",
    "    \n",
    "    system(glue::glue(\"\n",
    "    cd {working_dir}/Eagle_Phasing\n",
    "    {working_dir}/software/bcftools/bcftools isec -p chr{chr}_intersection -Ou chr{chr}.recoded.ccle.all.called.bcf {working_dir}/Eagle_Phasing/1000_genomes_vcfs/ALL.chr{chr}.phase3_integrated.20130502.genotypes.bcf\n",
    "    \"))\n",
    "      \n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have all of the intersections, organize everything a bit so that our life is a bit easier in the future. Delete all of the old files, move all of the intersection files into the same directory, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Loop through all of the directories and move the files that we need to the top-level directory\n",
    "chromosomes = seq(from = 22, to = 1, by = -1)\n",
    "for(chr in chromosomes){\n",
    "    system(glue::glue(\"\n",
    "    cp {working_dir}/Eagle_Phasing/chr{chr}_intersection/0002.bcf {working_dir}/Eagle_Phasing/chr{chr}_ccle.bcf\n",
    "    cp {working_dir}/Eagle_Phasing/chr{chr}_intersection/0003.bcf {working_dir}/Eagle_Phasing/chr{chr}_1kg.bcf\n",
    "\"))\n",
    "}\n",
    "\n",
    "#Delete the isec directories\n",
    "for(chr in chromosomes){\n",
    "    system(glue::glue(\"\n",
    "    rm -rf chr{chr}_intersection\n",
    "\"))\n",
    "}\n",
    "\n",
    "#Delete all of the vcf files that were used as inputs\n",
    "system(glue::glue(\"\n",
    "rm *ccle.all.called*\n",
    "\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Index all of the new bcf files\n",
    "system(glue::glue(\"\n",
    "\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "\n",
    "for bcffile in *.bcf\n",
    "do\n",
    "{working_dir}/software/bcftools/bcftools index -f $bcffile\n",
    "done\n",
    "\n",
    "\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Phase the CCLE data with Eagle2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Phase the ccle data with the 1kg reference panel."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Perform phasing\n",
    "chromosomes = seq(from = 22, to = 1, by = -1)\n",
    "\n",
    "for(chr in chromosomes){\n",
    "    \n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "{working_dir}/software/Eagle_v2.4.1/eagle --numThreads {num.threads} --geneticMapFile={working_dir}/software/Eagle_v2.4.1/tables/genetic_map_hg19_withX.txt --outPrefix ccle_chr{chr}_aftereagle --allowRefAltSwap --vcfTarget chr{chr}_ccle.bcf --vcfRef chr{chr}_1kg.bcf --vcfOutFormat u\n",
    "bgzip ccle_chr{chr}_aftereagle.bcf\n",
    "\")) \n",
    "}\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Index all of the phased genotype files\n",
    "\n",
    "system(glue::glue(\"\n",
    "\n",
    "cd {working_dir}/Eagle_Phasing\n",
    "\n",
    "for bcffile in *aftereagle.bcf.gz\n",
    "do\n",
    "{working_dir}/software/bcftools/bcftools index -f $bcffile\n",
    "done\n",
    "\n",
    "\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run RFMix "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run RFMix to calculate the ancestry fractions for all of the cell lines.\n",
    "\n",
    "We are running this on a big machine right now, but it looks like rfmix is only using one core, and really not that much memory. So in the future we should probabl try to modify the code to either take advantage of multithreading or just plan to run this on a smaller machine to save a bit of $. I don't think the phasing really needs that many CPUs either. It seemed to run pretty quickly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosomes = seq(from = 22, to = 1, by = -1)\n",
    "\n",
    "for(chr in chromosomes){\n",
    "    \n",
    "system(glue::glue(\"\n",
    "cd {working_dir}/Eagle_Phasing \n",
    "export PATH=$PATH:{working_dir}/software/bcftools\n",
    "{working_dir}/software/rfmix/rfmix/rfmix -f ccle_chr{chr}_aftereagle.bcf.gz -r {working_dir}/software/rfmix/all/ALL.chr{chr}.phase3_v5.shapeit2_mvncall_integrated.noSingleton.genotypes.vcf.gz -m {working_dir}/software/rfmix/1kg_sample_map -g {working_dir}/software/rfmix/1000-genomes-genetic-maps/interpolated_from_hapmap/chr{chr}_genetic_map -o {working_dir}/rfmix_output/chr{chr}.rfmix.output --crf-weight=30 --chromosome={chr}\n",
    "\"))\n",
    "    \n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
